examples/transactions.c

```diff
@@ -184,135 +154,137 @@ create_transactional_producer (const rd_kafka_topic_partition_t *rktpar) {
         if (error)
                 fatal_error("init_transactions()", error);
 
-
-        /* Begin a new transaction */
-        error = rd_kafka_begin_transaction(rk);
-        if (error)
-                fatal_error("begin_transaction()", error);
-
         return rk;
 }
 
 
 /**
- * @brief Abort the current transaction and destroy the producer.
+ * @brief Rewind consumer's consume position to the last committed offsets
+ *        for the current assignment.
  */
-static void destroy_transactional_producer (rd_kafka_t *rk) {
+static void rewind_consumer (rd_kafka_t *consumer) {
+        rd_kafka_topic_partition_list_t *offsets;
+        rd_kafka_resp_err_t err;
         rd_kafka_error_t *error;
+        int i;
 
-        fprintf(stdout, "%s: aborting transaction and terminating producer\n",
-                rd_kafka_name(rk));
+        /* Get committed offsets for the current assignment, if there
+         * is a current assignment. */
+        err = rd_kafka_assignment(consumer, &offsets);
+        if (err) {
+                fprintf(stderr, "No current assignment to rewind: %s\n",
+                        rd_kafka_err2str(err));
+                return;
+        }
 
-        /* Abort the current transaction, ignore any errors
-         * since we're terminating the producer anyway. */
-        error = rd_kafka_abort_transaction(rk, -1);
-        if (error) {
-                fprintf(stderr,
-                        "WARNING: Ignoring abort_transaction() error since "
-                        "producer is being destroyed: %s\n",
-                        rd_kafka_error_string(error));
-                rd_kafka_error_destroy(error);
+        if (offsets->cnt == 0) {
+                fprintf(stderr, "No current assignment to rewind\n");
+                rd_kafka_topic_partition_list_destroy(offsets);
+                return;
         }
 
-        rd_kafka_destroy(rk);
-}
+        /* Note: Timeout must be lower than max.poll.interval.ms */
+        err = rd_kafka_committed(consumer, offsets, 10*1000);
+        if (err)
+                fatal("Failed to acquire committed offsets: %s",
```

Comment:
you consider lots of things to be fatal. is this your recommendation for a real-world app?

it's what i'd do, reasoning: 1. these problems are unlikely 2. there are (probably?) some scenarios for which killing the app and have something restart it is the only thing you can do 3. you might as well just do this for a relatively wide range of issues to keep the code simple, which has the added benefit of making error handling logic errors less likely.

it would be good to get the error handling strategy / philosophy articulated well.

Reply:
Someone once said: "error out early", so I did.

The reason for this aggressive error handling is to keep the example simple, otherwise the error handling matrix would obscure the meat of what we're trying to demonstrate in this here example.

We could add comments in the actually-not-so-fatal-places why it happens, how it can be handled, etc, but ultimately we want rich error objects to tell this tale instead, which pretty much boils down to if an error is retriable or not.
This is not possible with the current API though.
A hacky alternative for the legacy APIs could be something like:
```c
        err = rd_kafka_committed(consumer, offsets, 10*1000);
        if (err) {
               if (rd_kafka_err_is_retriable(rd_kafka_committed, err))
                     goto retry;
                fatal("Failed to acquire committed offsets: %s",
̣             ..
         }

```

or something that creates a rich error_t* object from an API and an error code, like so:

```c
rd_kafka_error_t *rd_kafka_err2error (rd_kafka_api_function_ptr, rd_kafka_resp_err_t err);
```

Reply:
or a per-thread last-error object (we already have this for the legacy-legacy errno to resp_err_t conversion):
```c
rd_kafka_error_r *rd_kafka_get_last_error_object(rd_kafka_t *rk);
```

Reply:
All these alternatives are a bit funky in C, but the high-level bindings can hide it.